File: examples/helix_case/helixExample.cpp
#include "robotDescription.h"


extern ofstream logging_output_file;  // defined in main.cpp
/*
 * Helix Under Gravity Example
 *
 * Define your soft robot structure(s), boundary conditions,
 * custom external forces, and loggers in the function below.
 */

void get_robot_description(int argc, char** argv,
                           const shared_ptr<softRobots>& soft_robots,
                           const shared_ptr<forceContainer>& forces,
                           shared_ptr<worldLogger>& logger,
                           simParams& sim_params) {

    sim_params.dt = 5e-3;
    sim_params.sim_time = 10;
    sim_params.dtol = 1e-3;
    sim_params.render_scale = 5.0;
    sim_params.nis = IMPLICIT_MIDPOINT;

    // Read vertices describing helical shape
    vector<Vector3d> vertices;
    load_txt<Vector3d>("examples/helix_case/helix_configuration.txt", vertices);

    // Create the helix using custom config initializer
    double radius = 5e-3;
    double young_mod = 1e7;
    double density = 1273.52;
    double poisson = 0.5;
    soft_robots->addLimb(vertices, density, radius, young_mod, poisson);

    // Fix top end
    soft_robots->lockEdge(0, 0);

    // Add gravity
    Vector3d gravity_vec(0.0, 0.0, -9.8);
    forces->addForce(make_shared<gravityForce>(soft_robots, gravity_vec));

    // Set logger to record nodes
    string logfile_base = "log_files/helix";
    int logging_period = 1;
    logger = make_shared<rodNodeLogger>(logfile_base, convert_float_to_scientific_str(young_mod),
                                        logging_output_file, logging_period);
}

File: py_examples/helix_case/helix_example.py
"""
    Helix Under Gravity Example

    This is a pybind version of helix_example.cpp

"""
import numpy as np
import sys
import py_dismech
from pathlib import Path
from functools import partial

sim_manager = py_dismech.SimulationManager()

soft_robots = sim_manager.soft_robots
sim_params = sim_manager.sim_params
render_params = sim_manager.render_params
create_joint = sim_manager.soft_robots.createJoint
add_to_joint = sim_manager.soft_robots.addToJoint
add_force = sim_manager.forces.addForce

############################

sim_params.dt = 5e-3
sim_params.sim_time = 10
sim_params.dtol = 1e-3
sim_params.integrator = py_dismech.IMPLICIT_MIDPOINT

render_params.render_scale = 5.0
render_params.render_per = 5

# Read vertices describing helical shape from a file
vertices = np.loadtxt(
    Path(__file__).parents[2] /
    'cpp_examples/helix_case/helix_configuration.txt')

# Create the helix limb with custom configuration
radius = 5e-3
young_mod = 1e7
density = 1273.52
poisson = 0.5

add_limb = partial(sim_manager.soft_robots.addLimb,
                   rho=density,
                   rod_radius=radius,
                   youngs_modulus=young_mod,
                   poisson_ratio=poisson)

# Add the helical structure as a sequential series of vertices
add_limb(vertices)

# Fix the top end of the helix (locking the first node)
soft_robots.lockEdge(0, 0)

# Add gravity
gravity_force = py_dismech.GravityForce(soft_robots, np.array([0.0, 0.0,
                                                               -9.8]))
add_force(gravity_force)

# Initialize and run the simulation
sim_manager.initialize(sys.argv)
while not sim_manager.simulation_completed():
    sim_manager.step_simulation()
